### PR TITLE
Dynamo Packages Host filters

### PR DESCRIPTION
**Pull request by QilongTang**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

This PR adds a host filter right next to the traditional package sorting option. I will look at illustrating the host dependencies on each package entry as well but the mock up data is not working for me. Regardless of that, the filter and package filtering code is done. Currently it is a global setting which will affect searched results and sorted results or any combination.

PackageManagerFilter_v3

image

image

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@DynamoDS/dynamo 

### FYIs

@gregmarr

### CODE REVIEW

File: src/DynamoPackages/PackageManagerClient.cs [hunk 1]
```diff
@@ -89,6 +89,20 @@ internal IEnumerable<PackageHeader> ListAll()
...
+        /// <summary>
+        /// Make a call to Package Manager to get the 
+        /// supported hosts for package publishing and filtering
+        /// </summary>
+        internal IEnumerable<string> GetSupportedHosts()
+        {
+            return FailFunc.TryExecute(() =>
+            {
+                var hosts = new Hosts();
+                var hostsResponse = this.client.ExecuteAndDeserializeWithContent<List<String>>(hosts);
```
Comment (QilongTang) [inline src/DynamoPackages/PackageManagerClient.cs]: @pinzart Can you confirm this code from your example works? From my debugging the Greg Client call keeps giving me a failure thus a get an empty list

Comment (pinzart) [reply src/DynamoPackages/PackageManagerClient.cs]: @QilongTang Indeed there is an issue with the response format sent by PM server.
Fixing now...

Comment (pinzart) [reply src/DynamoPackages/PackageManagerClient.cs]: @QilongTang DEV should be fine now

Comment (QilongTang) [reply src/DynamoPackages/PackageManagerClient.cs]: @pinzart Thanks, will continue my work today

File: src/DynamoPackages/PackageUploadBuilder.cs [hunk 2]
```diff
@@ -52,7 +52,8 @@ public static PackageUploadRequestBody NewRequestBody(Package package)
 
             return new PackageUploadRequestBody(package.Name, package.VersionName, package.Description, package.Keywords, package.License, package.Contents, PackageManagerClient.PackageEngineName,
                                                          version, engineMetadata, package.Group, package.Dependencies,
-                                                         package.SiteUrl, package.RepositoryUrl, package.ContainsBinaries, package.NodeLibraries.Select(x => x.FullName));
+                                                         package.SiteUrl, package.RepositoryUrl, package.ContainsBinaries, 
+                                                         package.NodeLibraries.Select(x => x.FullName), new List<string>());
```
Comment (QilongTang) [inline src/DynamoPackages/PackageUploadBuilder.cs]: I can add a TODO here but pretty much this will be worked on for the package publishing part

**Comment by QilongTang** [comment]
Not sure why the AppVeyor build failed, the following code should build fine but throwing errors on that VM.
```C#
        public List<FilterEntry> HostFilter
        {
            get => hostFilter;
            set
            {
                hostFilter = value;
                RaisePropertyChanged("HostFilter");
            }
        }
```

**Comment by QilongTang** [comment]
It turns out the code above is in a newer .Net comes with VS 2019 which our build vm does not support yet, so I updated to use the traditional way.

**Comment by mjkkirschner** [comment]
@QilongTang if you want to use that you'll need to swtich to using msbuild 15 I think on the build machine. Our docker image should support this already, but I am not sure about the main pipeline (and very likely not self serve)

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchViewModel.cs [hunk 3]
```diff
@@ -294,6 +399,21 @@ public void Sort()
             this.SearchResults = temp;
         }
 
+        /// <summary>
+        /// Based on the known hosts received from back end,
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchViewModel.cs]: maybe specify the backend is the package manager server?

File: src/DynamoPackages/PackageManagerClient.cs [hunk 4]
```diff
@@ -89,6 +89,20 @@ internal IEnumerable<PackageHeader> ListAll()
...
+        /// supported hosts for package publishing and filtering
+        /// </summary>
+        internal IEnumerable<string> GetKnownHosts()
+        {
+            return FailFunc.TryExecute(() =>
+            {
+                var hosts = new Hosts();
+                var hostsResponse = this.client.ExecuteAndDeserializeWithContent<List<String>>(hosts);
+                return hostsResponse.content;
+            }, new List<string>());
```
Comment (mjkkirschner) [inline src/DynamoPackages/PackageManagerClient.cs]: does this return an empty list if this request fails? How long until it times out?

File: src/DynamoPackagesUI/DynamoPackagesUI.csproj [hunk 5]
```diff
@@ -33,8 +33,8 @@
     <WarningLevel>4</WarningLevel>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="Greg, Version=1.0.6176.18754, Culture=neutral, processorArchitecture=MSIL">
-      <HintPath>..\packages\Greg.1.0.6176.18754\lib\net45\Greg.dll</HintPath>
+    <Reference Include="Greg, Version=1.2.7121.21318, Culture=neutral, processorArchitecture=MSIL">
+      <HintPath>..\packages\Greg.1.2.7121.21318\lib\net45\Greg.dll</HintPath>
```
Comment (mjkkirschner) [inline src/DynamoPackagesUI/DynamoPackagesUI.csproj]: we should delete this project.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchViewModel.cs (hunk 3 above)
Comment (QilongTang) [reply src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchViewModel.cs]: Will do

File: src/DynamoPackages/PackageManagerClient.cs (hunk 4 above)
Comment (QilongTang) [reply src/DynamoPackages/PackageManagerClient.cs]: Yes, empty list if fails. Times out depends on the `HttpClient.Timeout` initialization in Greg, I think it is 100 sec without touching the `Timeoutspan` property.

File: src/DynamoPackagesUI/DynamoPackagesUI.csproj (hunk 5 above)
Comment (QilongTang) [reply src/DynamoPackagesUI/DynamoPackagesUI.csproj]: Totally, is this the js version pm client?

Comment (mjkkirschner) [reply src/DynamoPackagesUI/DynamoPackagesUI.csproj]: yes, we can remove this at any time.

Comment (QilongTang) [reply src/DynamoPackagesUI/DynamoPackagesUI.csproj]: I see, seems it is in the repo but not part of the solution, so little risk

File: src/DynamoCoreWpf/Views/PackageManager/PackageManagerSearchView.xaml [hunk 6]
```diff
@@ -173,6 +175,29 @@
 
                 </Button>
 
+                <Button Width="Auto" 
+                        Grid.Column="2" 
+                        Click="OnFilterButtonClicked" 
+                        fa:Awesome.Content="Filter"
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/PackageManager/PackageManagerSearchView.xaml]: What do you think of also adding a tooltip here? to the filter icon if it doesn't exist already?

File: src/DynamoCoreWpf/Views/PackageManager/PackageManagerSearchView.xaml [hunk 7]
```diff
@@ -262,10 +287,13 @@
                                         <StackPanel Orientation="Horizontal">
                                             <Button Style="{StaticResource FlatButton}" Command="{Binding ToggleIsExpandedCommand }" Background="#333">
 
-                                                <StackPanel Width="340" IsHitTestVisible="True" >
-                                                    <TextBlock FontSize="14"  Name="name" Text="{Binding Path=Model.Name}" Margin="5,0,0,0" HorizontalAlignment="Left" Foreground="WhiteSmoke" />
-
-                                                    <StackPanel Orientation="Horizontal" Margin="0,6,0,2">
+                                                <StackPanel Orientation="Vertical" Width="340" IsHitTestVisible="True" >
+                                                   <StackPanel Orientation="Horizontal">
+                                                      <TextBlock Width ="Auto" MinWidth="235" FontSize="14"  Name="name" Text="{Binding Path=Model.Name}" Margin="5,0,0,0" HorizontalAlignment="Left" Foreground="WhiteSmoke" />
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/PackageManager/PackageManagerSearchView.xaml]: what about a more specific name?

Comment (mjkkirschner) [reply src/DynamoCoreWpf/Views/PackageManager/PackageManagerSearchView.xaml]: whoops sorry, I see this is actually the name 😉 - thought it was the host.

File: src/DynamoCoreWpf/Views/PackageManager/PackageManagerSearchView.xaml (hunk 6 above)
Comment (QilongTang) [reply src/DynamoCoreWpf/Views/PackageManager/PackageManagerSearchView.xaml]: I can apply the same tooltip which I just added also to the context options, but the button I think need some different messaging?

File: src/DynamoPackages/Package.cs [hunk 8]
```diff
@@ -164,6 +164,7 @@ public static Package FromJson(string headerPath, ILogger logger)
                 if (body.name == null || body.version == null)
                     throw new Exception("The header is missing a name or version field.");
 
+                // TODO: Add serialization part of hosts specified in package.json
```
Comment (mjkkirschner) [inline src/DynamoPackages/Package.cs]: did you verify if this now gets serialized?

Comment (QilongTang) [reply src/DynamoPackages/Package.cs]: The property is serialized but it is empty. e.g.
```Json
{
    "license": "",
    "file_hash": null,
    "name": "tesingDepIssues",
    "version": "0.0.1",
    "description": "tesingDepIssues",
    "group": "",
    "keywords": null,
    "dependencies": [],
    "contents": "testingDepIssues - upload civil",
    "engine_version": "2.1.1.896",
    "engine": "dynamo",
    "engine_metadata": "",
    "site_url": "",
    "repository_url": "",
    "contains_binaries": false,
    "node_libraries": []
}
```

Comment (mjkkirschner) [reply src/DynamoPackages/Package.cs]: I don't see the `host` property there - `dependencies` is not the same property - it marks dependencies on other packages.

File: src/DynamoCoreWpf/Views/PackageManager/PackageManagerSearchView.xaml (hunk 6 above)
Comment (QilongTang) [reply src/DynamoCoreWpf/Views/PackageManager/PackageManagerSearchView.xaml]: @mjkkirschner More tooltips added

File: src/DynamoPackages/Package.cs (hunk 8 above)
Comment (QilongTang) [reply src/DynamoPackages/Package.cs]: Sorry I was confused by myself, so my TODO comment is valid then :)

**Comment by QilongTang** [comment]
@mjkkirschner Let me know if you prefer more changes on this one. Otherwise trying to get a real build by demo tomorrow

File: src/DynamoCoreWpf/Properties/Resources.resx [hunk 9]
```diff
@@ -2219,4 +2219,13 @@ Uninstall the following packages: {0}?</value>
...
     <value>Could not re-direct to the Dynamo forum page for feedback:</value>
   </data>
+  <data name="PackageHostDependencyTooltip" xml:space="preserve">
+    <value>This package requires the specified host(s) to run.</value>
+  </data>
+  <data name="PackageHostDependencyFilter" xml:space="preserve">
+    <value>Click to choose one or multiple host filter to browse host specific packages.</value>
+  </data>
+  <data name="PackageHostDependencyFilterContextItem" xml:space="preserve">
+    <value>Packages requires this specified host to run.</value>
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Properties/Resources.resx]: how about: 
`Packages requires the specified host to run.`

File: src/DynamoCoreWpf/Properties/Resources.resx [hunk 10]
```diff
@@ -2219,4 +2219,13 @@ Uninstall the following packages: {0}?</value>
   <data name="ProvideFeedbackError" xml:space="preserve">
     <value>Could not re-direct to the Dynamo forum page for feedback:</value>
   </data>
+  <data name="PackageHostDependencyTooltip" xml:space="preserve">
+    <value>This package requires the specified host(s) to run.</value>
+  </data>
+  <data name="PackageHostDependencyFilter" xml:space="preserve">
+    <value>Click to choose one or multiple host filter to browse host specific packages.</value>
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Properties/Resources.resx]: what about `filter packages by Dynamo host application` ?

**Comment by mjkkirschner** [comment]
@QilongTang LGTM - I had some alternative tooltip text for you to consider first though.

**Pull request merged**